api/src/main/java/org/pmiops/workbench/api/ClusterController.java

```diff
@@ -186,115 +183,97 @@
   public ResponseEntity<ClusterLocalizeResponse> localize(
-      String projectName, String clusterName, ClusterLocalizeRequest body) {
+      String workspaceNamespace, ClusterLocalizeRequest body) {
+    DbWorkspace dbWorkspace = lookupWorkspace(workspaceNamespace);
     workspaceService.enforceWorkspaceAccessLevel(
-        body.getWorkspaceNamespace(), body.getWorkspaceId(), WorkspaceAccessLevel.READER);
-    workspaceService.validateActiveBilling(body.getWorkspaceNamespace(), body.getWorkspaceId());
+        dbWorkspace.getWorkspaceNamespace(),
+        dbWorkspace.getFirecloudName(),
+        WorkspaceAccessLevel.WRITER);
+    workspaceService.validateActiveBilling(
+        dbWorkspace.getWorkspaceNamespace(), dbWorkspace.getFirecloudName());
 
-    FirecloudWorkspace fcWorkspace;
+    FirecloudWorkspace firecloudWorkspace;
     try {
-      fcWorkspace =
+      firecloudWorkspace =
           fireCloudService
-              .getWorkspace(body.getWorkspaceNamespace(), body.getWorkspaceId())
+              .getWorkspace(dbWorkspace.getWorkspaceNamespace(), dbWorkspace.getFirecloudName())
               .getWorkspace();
     } catch (NotFoundException e) {
       throw new NotFoundException(
           String.format(
               "workspace %s/%s not found or not accessible",
-              body.getWorkspaceNamespace(), body.getWorkspaceId()));
+              dbWorkspace.getWorkspaceNamespace(), dbWorkspace.getFirecloudName()));
     }
-    DbCdrVersion cdrVersion =
-        workspaceService
-            .getRequired(body.getWorkspaceNamespace(), body.getWorkspaceId())
-            .getCdrVersion();
+    DbCdrVersion cdrVersion = dbWorkspace.getCdrVersion();
 
     // For the common case where the notebook cluster matches the workspace
     // namespace, simply name the directory as the workspace ID; else we
     // include the namespace in the directory name to avoid possible conflicts
     // in workspace IDs.
-    String gcsNotebooksDir = "gs://" + fcWorkspace.getBucketName() + "/notebooks";
-    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
-    long workspaceId =
-        workspaceService
-            .getRequired(body.getWorkspaceNamespace(), body.getWorkspaceId())
-            .getWorkspaceId();
+    String gcsNotebooksDir = "gs://" + firecloudWorkspace.getBucketName() + "/notebooks";
+    long workspaceId = dbWorkspace.getWorkspaceId();
 
     body.getNotebookNames()
         .forEach(
             notebook ->
                 userRecentResourceService.updateNotebookEntry(
                     workspaceId, userProvider.get().getUserId(), gcsNotebooksDir + "/" + notebook));
-    String workspacePath = body.getWorkspaceId();
-    if (!projectName.equals(body.getWorkspaceNamespace())) {
-      workspacePath =
-          body.getWorkspaceNamespace()
-              + FireCloudService.WORKSPACE_DELIMITER
-              + body.getWorkspaceId();
-    }
 
+    String workspacePath = dbWorkspace.getFirecloudName();
```

Comment:
on ln 263, I was confused that `notebook` was the variable but it was appended to a string. `notebookName` might be better.

Also, it would be nice to have ` userRecentResourceService` expose a batch method for this, so you don't have to loop here. It's a bit odd to see a service getting called in a loop.